### DISCUSSION TITLE
Anytype API research: share your use-cases with devs

### DISCUSSION THREAD

**Discussion by vas3k** [anyproto/contributors#218] [Announcements]
Hey, everybody! My name is Vasily, I'm a new member of Anytype team. We've recently started a research about API implementation for Anytype. The feature that you all and we have been wanting for a long time!

Because of our p2p and e2ee architecture, as you can imagine, the API in Anytype is not just a matter of coding a couple of REST-endpoints on a "server". Things are going to be a little trickier in our lives :)

To better understand what we should focus on first and not make an “abstract API” that nobody can use, we want to collect some use-cases from you.

### Please tell us what you could use the API in Anytype for?

What bots, integrations, background services would you like to see first or code yourself? Maybe you've had an idea in your head for a useful integration for a long time, share it too!

### Rules:

- 1 top-level comment = 1 use-case. You can write several comments but please don't put them all in one, so that voting is cleaner.
- Vote for the use-cases! So we'll see which ones are more in demand.
- Feel free to discuss details, ideas and technical implementation in the 2nd level comments.
- No off-topic please. If you want to discuss APIs in general, please open another discussion.

### Template:

Use the template for better readability. We're going to refer to this thread a lot, so it would be nice to have it clean and organized.

```
**Short description of your use-case or idea**

More details about what you're trying to accomplish, how API integration would help you, what you've tried so far...
```

### Related discussions

- https://github.com/orgs/anyproto/discussions/210
- https://github.com/orgs/anyproto/discussions/211

**Comment by rushthomas** [comment]
Use Case: Triggering the creation of calendar events on Google Calendar when certain criteria are met. For example, if an object is Type:task and has a "Date" relation, create a google calendar event, titled with the Object Title.

**Reply by schmoelder** [reply]
While at it, also consider other protocols such as CalDAV, exchange etc.

**Comment by mishachepi** [comment]
**Creating an object with relations**

I think this is obviously the basic functionality that is needed. You need to be able to create an object with a specific type and specific relations and content.

**Comment by Onefabis** [comment]
Allow to save private channels from telegram with https://core.telegram.org/api/takeout into space where each post of that channel is object (note or page)

**Comment by alexjp** [comment]
**Update other relations when one relation changes**

Currently, I have tasks set up in a way, that when I click "DONE" checkbox on the layout, I have to go change the "Status" select option to "DONE" and set the "DONE Date" to today.

Would like an API that when I:

* Click Layout DONE from unchecked to checked
  * sets "Status" select option do "DONE"
  * sets "DONE DATE" to "Today"
* Click Layout DONE from checked to unchecked
  * sets "Status" select option do "TODO" (only if "Status" is on "DONE")
  * clears "DONE DATE"

* Click "Status" select option to "DONE"
  * Set Layout DONE to Checked
  * Set "DONE DATE" to Today
* Click "Status" select option to other option than "DONE"
  * Set Layout DONE to Unchecked
  * clears "DONE DATE"

* Click "DONE DATE" to Today
  * Set Layout DONE to Checked
  * sets "Status" select option do "DONE"
* Click "DONE DATE" to "empty/cleared"
  * Set Layout DONE to Unchecked
  * sets "Status" select option do "TODO" (only if "Status" is on "DONE")

**Reply by iBerlioz** [reply]
I believe that's not a part of API, but definitely the feature is really promising. I'd suggest you post it in feature requests

**Reply by nikkitavr** [reply]
buut.. if api will provide a kind of webhooks for events + basic crud "endpoints", devs could implement automations like these by his own

**Comment by wawax** [comment]
bulk import of data coming from a database into a collection.
creation of relation between objects using the API

**Reply by beaucronin** [reply]
bulk import is my dealbreaker -- if I can do this, including making the proper object relations, my usage of Anytype is completely transformed

**Comment by mishachepi** [comment]
**Get objects filtered by object_type and relation**

My use-case: I have an object "day" with relation "date". I need to get this object filtered by current date (my next steps: parse it -> send it to TG and mail in the morning).

**Comment by MikeyP888** [comment]
Use Anytype with Make.com for integration with other apps and automation. For example, I would want to use make.com to fetch data from anytype, then make.com would perform some action, then add a completed date or checkmark or change a single / multi select or tag in anytype. 

For example, let's say I want to use make.com and Anytype to help me write articles with ChatGPT. In Anytype, I might have a set of article titles with rough notes in the body of the object. I would use make.com to find the set, take the notes and title of the article, send it to ChatGPT, which would create an article . Then perhaps the completed article would be saved in Anytype ready to publish. Or it would be marked as complete and a date added to Anytype.

**Reply by mishachepi** [reply]
I have a similar use case.
There are many automation tools, n8n, Zapier, IFTTT. To use API to integrate with this service would be awesome.
To find many use cases you can check:
[Supported Actions paragraph for notion in n8n](https://n8n.io/integrations/notion/)
[Supported triggers and actions for notion in Zaiper](https://zapier.com/apps/notion/integrations?utm_source=google&utm_medium=cpc&utm_campaign=gaw-gbl-nua-search-brand-remarketing&utm_term=zapier%20notion&utm_content=9198877&utm_ads_campaign_id=17613359682&utm_ads_adset_id=139730148618&utm_ads_ad_id=608943269276&gad_source=1&gclid=CjwKCAjwko21BhAPEiwAwfaQCK1xyeW8xBpu-2qeXwK4NWw_d-3OSQAXbz82FvDN7Yf0IBfaSEuDlxoCK0IQAvD_BwE#triggers-and-actions)

**Reply by MikeyP888** [reply]
Yeah, I haven't tried n8n. My favourite is make.com. To the developers: if you do focus your efforts on these tools, please don't just focus on Zapier (like some apps do), it is super expensive!

**Reply by wawax** [reply]
there's also 2 new open source automation alternatives: actievepieces and automatisch

**Reply by andyjsiegel** [reply]
i highly recommend Pipedream. It's Make, but on steroids. Way more generous free plan and more flexible too.

**Reply by iBerlioz** [reply]
Totally support. n8n and etc compatibility would boost possible use cases

**Reply by FinnSour** [reply]
N8n is also able to be self hosted, which feels values aligned with anytype

**Comment by MikeyP888** [comment]
Another API use case I would like is using Anytype with Raycast (on Mac). For example, I would like to use it to create global keyboard shortcuts for things like creating a specific type of object or going to a specific page or set

**Reply by andyjsiegel** [reply]
Raycast Integration would be incredible, I definitely plan to work on that once the API is released. However, do note that you can click the 3 dot menu in the top right and click copy link, then make a quick link with it. Then, you can assign a keyboard shortcut to open a specific page already!
<img width="264" alt="image" src="https://github.com/user-attachments/assets/ebe21c74-3bba-4245-842a-4dbeed544657">

**Reply by jmetrikat** [reply]
@MikeyP888 @andyjsiegel Good news! The Raycast extension for Anytype is live 🎉

Grab it here: https://www.raycast.com/any/anytype or check out the repo: https://github.com/anyproto/anytype-raycast

**Comment by schmoelder** [comment]
**Sync Tasks with Github/Lab Issues / PRs**

When using Anytype for task management, it would be very helpful to have some connection with the issues / PRs  of certain Github/Lab repos that I'm working on. E.g. I would like to see them in my daily to do overview. See due dates, tags etc.

**Reply by avinash-singh-io** [reply]
It would be great if it supports Atlassian applications like jira, bitbucket, confluence

**Reply by typed-sigterm** [reply]
It would be great if it supports self-hosted Gitea too.

**Comment by schmoelder** [comment]
**Sync bookmarks with Firefox**

I'm aiming at converging all my knowledge in one place. So storing bookmarks in Anytype makes sense to me. However, I still need the bookmarks in my actual browser. Syncing them via an API would simplify this.

**Reply by avinash-singh-io** [reply]
not just only firefox but other browsers as well like chrome, edge, arc

**Reply by andersoal** [reply]
Maybe [floccuss](https://github.com/floccusAddon) can enable this

**Comment by Inuart** [comment]
An "Eat your own dogfood" API. That is, **build an API that the Anytype team is forced to use to rewrite the Desktop and mobile apps**.

There should be no other way to interact with the Go core. **No internal APIs**.

That's the best way to ensure we get an API that's fully featured and of great quality for everyone.

**Reply by vas3k** [reply]
That's our plan too :) A proper API will allow us to develop side-plugins, importers and other integrations without disturbing the core-team.

**Reply by Inuart** [reply]
Just the plugins? I meant to say "use the API for everything". Like, rebuild the desktop and phone apps using only the API to interact with the Go core.

I should rephrase my comment to make it clearer.

**Reply by vas3k** [reply]
Noted. That's why we're asking for actual use-cases

**Comment by Self-Perfection** [comment]
**Use anytype as a backend for pomodoro/time tracking widget**

I imagine something like reimplementing of https://intend.do/ on top of Anytype. With desktop widget that allows to:

1. Select one tasks from today plan as active or create new
2. View how much time is left in my current working session (like pomodoro timer)
3. Stores in each task how much time/pomodoros were spent on this task
4. Let group tasks by projects (goals in intend.do terms) and gather aggregated stats on how much time were spent on each task for longer time durations, like weeks or months.

**Reply by andersoal** [reply]
A task/calendar/habits builder with streak, pomodoro, session (countdown timers), timetracker … there is indeed a bunch of features to explorer

**Comment by ChristianHaake** [comment]
Hi! 

using low / no-code solutions to automate monkey work:
- create / modify objects and content
- connect different applications 

I'm not a coder, but I would need a way to add Links to folder/files in nextcloud or o365. the view need to stay synced. 
The goal is to have a one contextual view to have all the information in one place. My notes etc as well as the files

**Reply by iBerlioz** [reply]
n8n/make.com solutions mentioned above

**Comment by andyjsiegel** [comment]
Integrating Anytype into Tasker, Raycast, and Pipedream would be my top 3 so I can integrate data from external sources.

**Comment by mishachepi** [comment]
**Spaced repetition of specific notes. Like Anki.**

The implementation will be carried out using some third-party tool, for example: no-code tool n8n (if script execution is not added to Anytype in the future). From Anytype API, it is required to receive objects by tag, and change the relation field with date of the last repetition.

**Comment by lazjedi** [comment]
**Google Drive and other cloud services integration**

As in Notion, we should be able to add files in Google Drive or other cloud service as embeds to our pages.

Like this;
![Screenshot_20240819-122613](https://github.com/user-attachments/assets/68e121e3-520f-45b8-910e-fc95c2e116db)

**Comment by ebanDev** [comment]
**Quickly capture infos**

I already created a software using the API that allows me to quickly capture thoughts or literally anything to AnyType from CLI I just publised it here : https://github.com/ebanDev/AnytypeCapture

Hope that helps ! :)

**Comment by sreeharsha-rav** [comment]
There is going to be a lot of local data, so I want to develop custom AI features with the help of local LLMs, if Anytype is not going to allow integration for Local LLMs in the future, it's not something I would expect to use long term.

**Reply by algusarov** [reply]
Totally support, wanted to add the same - connecting personal productivity tools which are GenAI/LLM-based with Anytype as knowledge repo / knowledge graph. Examples: MCP to interact with Anytype objects (query your favorite movies, query pending high prio Todo - via MCP-supported AI tools like Claude Desktop in natural language). For MCP few implementations exist, but they all seem to connect to Desktop, which is very limiting (can't trigger actions from mobile device for example). Also OpenAI assistant API, tool calling, building agents for PKM, building RAG/GraphRAG for Anytype data. Anytype has a super strong edge over other PKM-backend for such use cases as there is more metadata for objects available, so LLM processing can provide much better outputs by querying and processing right types.

**Comment by ZMensRain** [comment]
### Use case: 

Publishing new html documents to a website when new pages and notes are created.

### Further information

I would like to be able to create an object (mainly pages and notes) and mark it with a relation as public or private (that is pretty easy) then I would like to automatically add new hyperlinks and pages to my website based on the objects. I could see this being pretty useful for people who write in the personal knowledge base and want to share their writings on the web

**Comment by andscape-dev** [comment]
### Bulk modifications

I often find myself discovering a new way to enrich or structure my objects, and wanting to apply an update to some/all objects of a type. As a dev, with a relatively easy to use API I could script my own bulk modifications, but even better would be to have official or community built tools for it using that API.

**Comment by iBerlioz** [comment]
Nextcloud integration
For instance: a property for an object to attach a file, which will be uploaded to a Nextcloud instance to certain folder. This way it will be possible to easily share files and keep them organised in a convenient way

**Comment by avinash-singh-io** [comment]
Google Sheet Export and Import
We should be able to connect with Google sheet, if we create a daily tracker calendar then we should be able to export it for analysis or we should be able to analyze it in the app itself

**Comment by nikkitavr** [comment]
**At least - just basics crud operations for core domain models**

Idk what to say else, it's basics

**Comment by nikkitavr** [comment]
**A kind of webhooks or full pub/sub model for domain events**

For a really big opportunities of customization, a classic api client-server paradigm is not the best choice

Im not a spec in a p2p communication, but maybe if you don't have a classic server, it's more flexibility to implement publish/subscribe model for basic domain events e.g "object with id change state"

**Reply by RawmanGrey** [reply]
> **A kind of webhooks ... **


**Non-blocking POST request with structured data on user-selected relations change**

It is all about webhooks/triggers/notifications... you name it. 
I imagine an option to create a relation in UI which would take following properties:
1) URL to make POST request
2) Headers for this request
3) List of other relations to subscribe
4) Include *onlySelected*  or *allData* toggle

Request body should contain current and previous data state, reflecting option 4 schema.
Pushes on subscribed object deletion seem easy, but pushes on creation seem vague: relevant data will be changed later.  

Having an option to trigger events outside of Anytype is a must to my mind for collaborative project management.

If some [mustache](https://github.com/mustache/mustache.github.com) syntax could be used to refer changed values in body/headers, such interface could communicate with services like [ntfy.sh](https://ntfy.sh/) without any other intermediate step. 


PS:
Storing "secrets" somewhere in space options as a globally available variable would be nice, just not to have them on the surface : )

**Comment by travellingtechie** [comment]
I'm not sure if this would qualify as an API or something different, but I'd like to be able to interact with my content as an unencrypted  filesystem (similar to cyrptomator).  I currently use obsidian and one of the things that I like is that it's just a bunch of md files in a directory, I can open them up in vscodium, I can run python scripts against them, I can trigger other actions when a file gets updated in a directory.  I think that in itself would cover a lot of the API use cases, if not as cleanly as an API.

As far as how API access would work, my first thought was that it would be an API node type running on a server, similar to consensus and coordinator nodes.  Although it would be nice to also be able to run an API locally, being able to access it as an unencrypted local filesystem  would serve most local purposes.

**Comment by rrmillar** [comment]
Use Case: Replacing GDocs, GSheets and Google Keep for personal and project documentation.

**Comment by dustinosity** [comment]
I'd like to be able to hit the API from an automation inside Home Assistant to create new objects. Specifically creating new notes or tasks via the HA's new voice assist pipeline that handles the text-to-speech, does intent matching, and can trigger the appropriate automations.

**Reply by dustinosity** [reply]
To get local api access, I'd be fine running an always-on install of anytype in a container on the same network as the Home Assistant install, though a headless edition of the client that I could run in docker or lxc would be amazing. I could also just use tailscale to get easy secure access to the api from anywhere.

**Comment by Breezy-Breeze** [comment]
Vas3k what is the team's policy on social media content creators? I am interested in starting a hobbyist channel around this and don't want to step on any toes. Thanks.

**Comment by mishachepi** [comment]
**bulk action automatization**
I have a lot of objects and make some big changes sometimes is difficult for standard client
- Replace all links to Object1 to Object2 link in whole space
- Add specific relation to all specified objects
- Transfer relation value from Relation1 to Relation2 in specified object
...

**Comment by zinkkrysty** [comment]
**Make a Chrome/browser extension or mobile app for saving links & managing bookmarks**

Something like Pocket or Delicious (which are either shut down or about to shut down), but built on top of AnyType, so the user is in control of their data.

There should be a way to save links, add notes and organize bookmarks in a folder-lile structure. Essentially I also want to replace Arc (pinning & organizing tabs) which was discontinued with this extension.

The API integration should allow to create and manage objects (links+notes) and to be able to categorize/tag them.

The AnyType sync engine would allow to use this extension/app from any of my devices.

**Comment by zy55769068** [comment]
**Short description of your use-case or idea**  
Expose full rich text formatting information—such as callout styles, text/background colors, and code block language—**both when reading and writing through the API**, so that MCP-based or other apps can fully preserve, render, and manipulate content accurately.

**More details**  
Now that the MCP server is available, it’s much easier to build tools and integrations that interact with Anytype. However, since MCP operates on top of the underlying API, its capabilities are inherently limited by what the API exposes.

Currently, the content returned by the API (and thus via MCP) is mostly simplified into **basic Markdown**, which strips away much of the visual and structural richness of notes created inside Anytype.

This is not only an issue when **reading/exporting** content via MCP or API—it also limits our ability to **write rich content back** into Anytype through these interfaces.

Here are some specific formatting limitations:

- **Callout blocks**: No metadata about the type (info, warning, etc.), associated emoji/icon, or background color.
- **Text formatting**: Inline text colors, highlights, and background colors are lost.
- **Code blocks**:
  - When **creating pages via the API**, only a few languages are recognized and rendered correctly in the app (e.g. `javascript`, `sql`, `java`, `go`, `rust`, `php`).
  - Other languages like `python`, `c#`, `typescript`, or `ruby` are **ignored or silently downgraded**.
  - Even worse, when **reading existing pages**, all code block language annotations are **completely stripped out**, so it's impossible to preserve the original syntax highlighting.
- **Advanced block types**: Toggles, layout blocks, or object-style links are flattened or omitted.

This makes it difficult to:

- Build high-fidelity viewers or renderers that preserve original formatting.
- Create MCP-based tools that **transform, generate, or sync** content while keeping its structure and style.
- Import or generate notes programmatically with accurate code snippets and formatting.
- Develop backup/export workflows that retain visual and semantic richness.

**Suggestion**  
Please consider enhancing the API to expose a richer, structured format for note content (e.g. block-based JSON with formatting metadata), and also **accept this structure when writing content** back into Anytype.

Specifically:

- Code blocks should preserve their language annotations during both read and write.
- The list of supported languages should be expanded and made consistent between creation and rendering.
- A `fullRichFormat = true` or similar flag would be useful to access the complete formatting structure for reading.

This would enable the community to build powerful and visually accurate tools on top of Anytype's ecosystem—not just viewers or exporters, but also generators, importers, sync tools, and intelligent assistants.

Thanks again for your hard work and openness to feedback.

**Comment by tribals** [comment]
Use-case: automate retrieval of information about Books from online libraries (metadata).

Details: I want to create specific type for Books where I'm not forced to fill all the properties by hand. Instead, whenever I create an object of type "Book", and filling designated property (like "Title"), then it triggers external helper that can interact with online libraries in order to find metadata about book, like author, year published, edition, etc. An external helper is completely transparent to AnyType, and interacts solely through API. On the other hand, it can speak with many online book databases (configured) and ideally retrieve metadata in any desired language, not only one.

**Comment by ArsalaBangash** [comment]
**Trouble Getting MCP server to add properties though MD content worked during creation**

  I've been collaborating with an AI assistant running on the anytype-mcp server to 
  import my music release workflow from a CSV file into Anytype. The goal was to 
  create a "Song" object for each row in the CSV, with the object's properties and 
  content populated from the CSV data.

  This turned into a deep dive into the Anytype ecosystem. We started by trying to 
  use the local API, and we quickly ran into issues with creating and updating 
  objects with content. The API would return success, but the content would not 
  appear in the client, or it would be appended instead of replaced.

**Reply by jmetrikat** [reply]
@arsalabangash you can set `properties` during object creation through the API, as well as update them by patching an existing object. please refer to the documentation of endpoints on the developer portal: https://developers.anytype.io/docs/reference/2025-05-20/create-object

patching the blocks currently returned as markdown through the API will be introduced in the next update; as of right now it's not yet supported when updating an object

**Reply by ArsalaBangash** [reply]
Thank you @jmetrikat

**Comment by ADITHYA31** [comment]
@jmetrikat How do I get started if I want to contribute to this really amazing project, Iam a rookie Dev but confidant can catch up pretty quick.

**Reply by jmetrikat** [reply]
hey @adithya31 :)

you can find the main implementation of the api in the anytype-heart repository. before diving in, please note the significant advancements and fixes included in this pr:
- https://github.com/anyproto/anytype-heart/tree/develop/core/api
- https://github.com/anyproto/anytype-heart/pull/2448

the developer portal is linked below, and you can find the portal itself along with the specifications in the anytype-api repository:
- https://developers.anytype.io/
- https://github.com/anyproto/anytype-api

also make sure to check out the existing issues & feature requests here:
- https://github.com/anyproto/anytype-api/issues
- https://community.anytype.io/search?q=api%20order:latest

**Comment by TomTom101** [comment]
An API wrapper does not make a useful MCP server. Example: A search query with one result returns 1,610 tokens. The actual text containing the search keyword comprises just 33 tokens. That's 98% noise.

**Comment by dev-inside** [comment]
I am working currently on my Video Production Pipeline which is a selfhosted and pretty complex (but cool) task.

It works as follows:
**1. Create the project in n8n**
1. Creates the project by creating the folder structure in Nextcloud including a readme.md
2. Creates the project in Anytype via the CLI
3. If it has a deadline, it will create a calendar entry in my Caldav-Calendar
4. Sends notifications via Mail and ntfy.sh with the link to Nextcloud

**2. Work on the Project in Anytype Desktop**
1. The a project contains Collections and Tasks and render  tasks in a Kanban-Board after following scheme:
  -  Projects (Collection)
  -  The Project itself (custom Collection)
  -  A custom Task
2. I finish the first tasks like getting assets, start planing etc
3. One of the tasks contains to write a video script which is also the baseline for the storyboard and important for some automations later

**3. Create the shooting plan with n8n and Ollama**
- The next Automation will request the Project and parse the video-script to n8n. 
- It breaks down the chapters and creates Tasks for the shooting plan via Ollama
- Those tasks will be iterated again to create shooting tasks for B-Roll (detail shots for example for a product-review)

**4. Shooting Dashboard Home Assistant**
<img width="2254" height="1565" alt="image" src="https://github.com/user-attachments/assets/22231508-ef15-4a79-940d-9fd84b4c64e8" />

I managed to fully automate my hardware in my studio and can control everything via the Home Assistant dashboard. I use the following hardware and apps as follows:

### Teleprompter
**Hardware** Xiaomi MI A1 flashed with /e/os/
**Apps:**
- Freekiosk (Open Source): to have some control over the used apps via MQTT
https://github.com/RushB-fr/freekiosk
- Tiefprompt  (Open Source): as Teleprompter
https://github.com/Tiefseetauchner/tiefprompt
- Screenstream  (Open Source): to show a RTSP-Stream (Integrated via go2RTC)
https://github.com/dkrivoruchko/ScreenStream
- BT Cam (paid) to start/stop my cameras (Has a REST API)
https://www.btcam.app/

### Virtual Keyboard 
The remote control in Freekiosk seems not to work properly. So I needed a device which acts like a bluetooth keyboard. It connects via BT to my old phone.
**Hardware** ESP32-S3
**Firmware**:  via ESPHome https://github.com/dmamontov/esphome-blekeyboard

### Timecode Display
I needed a visual Timecode display, but they are expensive AF. So I build myself one. I'd ordered a Waveshare EPS32-S3 (https://www.waveshare.com/esp32-s3-touch-lcd-3.49.htm) device and tinkered something out:
![timecode](https://github.com/user-attachments/assets/a9f41cd0-df1d-4744-9051-db9b0483b7e0)

### Clapperboard
I bought an ESP32 E-Ink device from Seed-Studio (https://www.seeedstudio.com/reTerminal-E1001-p-6534.html) and created a Clapperboard for that as well.
![clapperboard](https://github.com/user-attachments/assets/c1d3658a-cb5b-42d6-86b3-7c12564cce65)

The Clapperboard gets the infos via the Anytype API and is fired via Webhook. It parses the Scene names to Home Assistant. I can then just switch via a dropdown the scene name. No more tiping!

The Clapperboard and the Timecode display are fully controlled via the dashboard. The Scene and Project name come from the Shooting plan list and I can change the takenumber by pressing buttons on the device.

I know this is a lot, but this is insanely efficient and saves me a lot of copy paste and hours in creating tasks and also a lot of time on set. The Timecode Display and Clapperboard are extremly cool, but I will replace them wit a smaller device in the future. For the shooting, I need actually only the Scene Name, the take and the timecode and I found on the waveshare website the fitting device:

![RLCD](https://github.com/user-attachments/assets/f58f0f00-d156-47ae-8e7d-c5626cfdade8)
https://www.waveshare.com/esp32-s3-rlcd-4.2.htm?sku=33298 

If you have any questions, let me know! And if you made it to the end: thank you!

**Reply by dev-inside** [reply]
So after that long introduction:
I don't miss too much in the API, but I would love to have better documentation and examples. I am figuring out, how I can add to a certain collection a new Object. This is not really clear to me at the moment.

**Reply by jmetrikat** [reply]
> So after that long introduction:
> 
> I don't miss too much in the API, but I would love to have better documentation and examples. I am figuring out, how I can add to a certain collection a new Object. This is not really clear to me at the moment.

@dev-inside add an existing object to a collection via: https://developers.anytype.io/docs/reference/2025-11-08/add-list-objects

to create a new object before, use: https://developers.anytype.io/docs/reference/2025-11-08/create-object

**Reply by dev-inside** [reply]
@jmetrikat 
After tinkering around with the API, it would be really great to add the object directly to the list (optional of course) via a list ID. I am testing it at the moment with the n8n-node and it creates an object, but the node doesn't return a response. 

This means that I have to make after the "create object" an another call to find my object, then the last one to add it to the collection. This is extremely inconvenient. I mean, why allow people adding templates, but not where they are supposed to be used?! :wink: 

Don't know if the missing response is an error of the node, haven't tested that call so far via traditional REST (will tomorrow).

**Reply by dev-inside** [reply]
@jmetrikat 
I played around with templates with collections. The items are only linked and not copied, this is unfortunate cause I would love to create a collection with predefined Tasks which are bound to the Project-Collection.

The second thing is (but this is UI-Related), if I delete a collection only the collection is deleted and not the objects inside of it.

<img width="1963" height="1102" alt="image" src="https://github.com/user-attachments/assets/dbc07399-a9bb-4495-8517-2c227dd00676" />

I can of course create them via API, but IMHO this is kinda a design flaw of Anytype.

**Reply by jmetrikat** [reply]
> After tinkering around with the API, it would be really great to add the object directly to the list (optional of course) via a list ID. I am testing it at the moment with the n8n-node and it creates an object, but the node doesn't return a response. This means that I have to make after the "create object" an another call to find my object, then the last one to add it to the collection. This is extremely inconvenient.

@dev-inside posting to `/v1/spaces/:space_id/objects` is gonna return the created object as response - including its id - no need to do another request to search it in 2nd request. might be a shortcoming of the n8n node otherwise.

> The second thing is (but this is UI-Related), if I delete a collection only the collection is deleted and not the objects inside of it.

this is the intended behavior and unrelated to the api endpoints discussed here

**Reply by dev-inside** [reply]
@jmetrikat 
> posting to /v1/spaces/:space_id/objects is gonna return the created object as respons...

Well, then I would assume it is user error on my side or the n8n-node has an issue. It creates the object, but I get an empty response.

> this is the intended behavior and unrelated to the api endpoints discussed here

I know and I already pointed it out ;) It's abosulutetly logical how Anytype is intended to be used, but I should maybe open an FR-Request on the Frontend-Repo (Guess TS is right) for that option.

**Comment by kinthaiofficial** [comment]
A few high-value API use cases from our experience running multi-agent systems:

**1. Knowledge graph traversal**: Anytype's graph structure is perfect for agent memory. Agents need to store and retrieve structured knowledge (entities + relationships), not just flat text. An API that exposes the graph natively — not just as a document list — lets agents build proper episodic memory on top of Anytype.

**2. Per-agent workspaces with isolation**: Each AI agent gets its own Anytype space, but the orchestrator can read across spaces (with permission). This maps naturally to multi-tenant agent architectures where each agent maintains its own knowledge base but can share specific objects with other agents.

**3. Event streaming for agent triggers**: When a human edits an object, agents watching that object get notified. This is the foundation for "human-in-the-loop" workflows — the human edits a document, the agent detects the change, processes it, and creates new objects with the result.

**4. Typed object schemas for structured agent outputs**: Agents produce JSON; Anytype should be able to store it as a typed object, not just a text blob. If an agent produces a structured analysis, it should become a proper Anytype relation set that humans can sort, filter, and query.

The biggest win for agents would be a structured query API (beyond text search) — agents need to ask "give me all Tasks modified in the last 7 days by any agent with priority = high," not just keyword search.

Writing about agent memory architecture: https://blog.kinthai.ai/why-character-ai-forgets-you-persistent-memory-architecture

**Comment by zhinn0** [comment]
Using AI to build out a wiki.

Seems to be an issue due to lack of block-level API? I can't use AI to create - using mcp & api - an inline pill that works that same as when I, in the editor, manualy use "@" or "/" and select the object. Perhaps I'm missing something?

More context:

Discussion #218 mentions that "patching the blocks currently returned as markdown through the API will be introduced in the next update."
Repro:
Read an existing object that contains an inline pill (created in the editor via "@" or "/"). The body comes back as:

   [Object Name](anytype://object?objectId=<id>&spaceId=<id>)

Write that exact same markdown back via update-object.
Observed result:

The link renders as visually similar plain underlined text in the editor
No icon is pulled from the linked object (real pills show the target's icon)
The source object's links system property stays null
The target object's backlinks system property stays null
Link does not appear in the graph view

Expected result: same behavior as "@" or "/" in the editor — full pill with icon, registered in links/backlinks, visible in graph.

Use case: programmatically building wikis with inter-linked pages. This is straightforward through the Notion API; the gap here is the only blocker for parity.
Tested formats that all fail: [Name](anytype://object?...), [Name](bareObjectId), @Name, [[Name]]

**Reply by staus** [reply]
Echoing @zhinn0, I run into issues getting object mentions to work. 

The only thing that gets close is linking using this syntax:
`[Name](anytype://object?objectId=...&spaceId=...)`

But as @zhinn0 mentions this is not treated as a real mention and doesn't create backlinks.

If not for this issue, I'm very happy and impressed by the current API, so well done!

**Reply by StringAbaper** [reply]
I hit the same wall and found a workaround that gives you real mentions, not just a parallel relation.

**Short version:** the object body doesn't create mentions as you both stated. Relations are declared through *properties*, and any custom property of format `objects` works. Anytype derives the system relations from it, both ways.

**What I tested** (local API `2025-11-08`, via `@anyproto/anytype-mcp`):

1. Created a custom property, format `objects`.
2. Set it on a page through `POST /objects` (or `PATCH`), pointing at another page:

   ```json
   "properties": [
     { "key": "my_property", "objects": ["<target_object_id>"] }
   ]

3. Result:
- the source object's links system property is populated
- the target object's backlinks system property is populated

**Practical note:** the body link and the property are independent, you have to set both.

**Comment by Antheurus** [comment]
Sharing a use-case: I built [anywrite](https://github.com/Antheurus/anywrite), a CLI covering all 52 endpoints of the local API (2025-11-08 spec) — spaces, objects, properties, tags, types, templates, lists, chat, files, members, search, auth. It's a Bun/TypeScript project compiled to a single dependency-free binary, with the whole endpoint surface encoded as a data registry rather than hand-written per-endpoint code.

The motivator was agent-driven usage specifically: the official MCP server loads all 52 tools into every session's context whether they're used or not, which is expensive for agents that only touch a handful of endpoints per task. anywrite is instead wired as a Claude Code skill — a normal CLI that costs zero context until an agent actually invokes it — while staying just as usable from a plain terminal.

A few things the live API surprised me with while building this, in case useful to others hitting the same spec: DELETE is an idempotent soft-archive on every resource type (repeat-delete stays 200, never hits the spec's documented 410 — that only fires on GET of an already-purged resource, which isn't reachable through any exposed DELETE call), file upload dedupes by content hash (re-uploading identical bytes returns the existing object's id rather than creating a new one), and unknown-id GETs return HTTP 500 rather than 404.

MIT-licensed, open source: https://github.com/Antheurus/anywrite
